text/0648-aspects-priority-ordering.md

```diff
@@ -0,0 +1,525 @@
+  the full ability to control their aspect invocation ordering.
+* Existing users would need to modify their code to adapt to this new encoding of aspects.
+
+#### 2. Multiple Passes of the Construct Tree with a Stabilization Loop
+
+We can perform multiple passes over the construct tree to ensure that if an Aspect creates a new node, that node will inherit its parent’s Aspects and
+be visited. To prevent infinite recursion, we can set a maximum iteration depth (e.g., 100 passes). If invokeAspects hits this limit while still mutating
+the tree, it will throw an error. View pseudocode for this alternate solution in the Appendix.
+
+Pros:
+
+* Solves the first constraint without requiring changes from the customer. Newly created nodes will automatically inherit and apply their parent’s Aspects.
+
+Cons:
+
+* This solution doesn’t address the third constraint of Aspect ordering. We would still need a way to specify the order for cases where mutating and
+validation Aspects apply to the same node.
+* Validation aspects may still run before mutating ones, causing them to throw Errors and fail when they shouldn’t.
+* * This solution will change customer’s infrastructure.
+    * This reason alone is enough not to pursue this solution. A solution that will change customer’s infrastructure without them making any changes
+    to their CDK code would be considered breaking.
+* Additionally, this solution would be difficult to understand for customers, compared to assigning priority values.
+  * Some times it will work as expected, but other times it will not (since we still have no way to prioritize the ordering of how mutating Aspects are
+  applied).
+
+### What are the drawbacks of this solution?
+
+The primary risk is that users may not immediately understand the need to specify priorities, leading to issues if aspects are applied in the wrong
+order. Additionally, this change may require users to revisit and update their existing code to take advantage of the new ordering system.
+
+### What is the high-level project plan?
+
+1. Submit this RFC and watch for feedback.
+2. Implement initial changes as described in this RFC.
+3. Implement additional changes (like potentially adding `priority` to the Aspect interface) depending on the feedback on this RFC.
+
+### Are there any open issues that need to be addressed later?
+
+We have discussed whether or not to add the `priority` property to the `IAspect` interface for users to set defaults.
+For now, we are not but we can discuss this in the RFC process.
+
+## Appendix
+
+### New Invoke Aspects Algorithm (Pseudocode)
+
+```
+function invoke_aspects(root):
+  aspects_map = collect_all_aspects(root)
+  
+  // Iterate through aspects sorted on priority:
+  for prio in aspects_map.keys().sorted():
+    for (aspect, node) in aspect_map[prio]:
+        invoke_aspect(node, aspect)
```

Comment:
Aspects are not collected for nodes added dynamically by other Aspects.
The collection of the aspects happens at the beginning before invoking them, if one of the "write aspects" adds nodes to the tree, they won't be collected and the aspects they should inherit would not be processes. This would not solve issue [21341](https://github.com/aws/aws-cdk/issues/21341)

would you add a loop to repeat the collection after the a pass of `invoke_aspects` you would end up with something similar to alternate solution # 2, however the processing order of the aspects would be inconsistent with the priorities.

a solution # 3 could be an hybrid between # 1 and # 2, where you:

1. Collect only priorities for all aspects on the tree
2. Pick the first priority and invoke all aspects on the tree for this priority that have not been processed before
3. Repeat from 1. until no new aspects are processed
4. Pick the second priority and do the same as above and repeat for all other priorities

However also for this solution there is a gap, if a dynamically added node has an Aspect with an higher execution priority, the above solution would miss it or execute it later than expected (if you re-gather the list of priorities)

I fear that the only solution is # 2, and the priority based execution order can be defined only as being valid within the scope of the node and not of the entire tree.